File: scripting-languages-winter-2017/perl/registry.pl
#!/usr/bin/perl
# Witold Bobrowski ; Group 2
#
# Parse text file of students with grades and generate registry

use strict;
use warnings;
use Utils;
use File::Basename;

# VARIABLES

my $longest_name_length = 0;
my $longest_grade       = 0;
my $longest_list        = 0;

# SUBROUTINES

# Parse file contents into a hash table
# Arguments:
#   1) string - file name
sub evaluate_inout_file {
    my %registry;
    open( my $data, '<:encoding(UTF-8)', $_[0] )
      or die "Could not open file '$_[0]' $!";
    while ( my $line = <$data> ) {
        $line =~ s/[\r\n]+$//;
        my @info       = split / /, $line;
        my $first_name = lc "$info[0]";
        my $last_name  = lc "$info[1]";
        compare_to_longest_name("$last_name $first_name");
        my $grade = Utils::get_grade( $info[2] );
        if ( $grade == 0 ) {
            print STDERR
"WARNING! \"$info[2]\" is not a valid grade, therefore it will be ignored\n";
        }
        else {
            compare_to_longest_grade($grade);
            push @{ $registry{"$last_name-$first_name"} }, $grade;
        }
    }
    return \%registry;
}

# Format array of student names into a one string prefixed with approproate amount of spaces
# Arguments:
#   1) string array - Array of names [$first_name, $last_name]
sub format_name {
    my @names      = split /-/, "$_[0]";
    my $last_name  = ucfirst $names[0];
    my $first_name = ucfirst $names[1];
    my $result     = "$last_name $first_name";
    my $n          = $longest_name_length - length($result);
    my $spaces     = ( ' ' x $n );
    return "  $spaces$result";
}

# Format grade by prefixing with approproate amount of spaces
# Arguments:
#   1) float - grade
sub format_grade {
    my $n      = $longest_grade - length("$_[0]");
    my $spaces = ( ' ' x $n );
    return "$spaces$_[0]";
}

# Compare given full name with the longest yet recorded, replace if it's longer
# Arguments:
#   1) string - fullname
sub compare_to_longest_name {
    for my $string (@_) {
        if ( length($string) > $longest_name_length ) {
            $longest_name_length = length($string);
        }
    }
}

# Compare given grade with the longest yet recorded, replace if it's longer
# Arguments:
#   1) string - grade
sub compare_to_longest_grade {
    for my $string (@_) {
        if ( length($string) > $longest_grade ) {
            $longest_grade = length($string);
        }
    }
}

# Create and return string with sorted students with their grades and average
# Arguments:
#   1) hash array reference - dictionary of grades for each students
#   2) string - file name
sub generate_output_files {
    my %registry               = %{ $_[0] };
    my $file_name              = basename( $_[1] );
    my $output                 = "";
    my $total_sum              = 0;
    my $total_number_of_grades = 0;
    foreach my $student ( sort keys %registry ) {
        my $sum              = 0;
        my $number_of_grades = 0;
        my $name             = format_name($student);
        $output = "$output$name    [ ";
        foreach my $grade ( @{ $registry{$student} } ) {
            my $formated = format_grade($grade);
            $output = "$output$formated, ";
            $sum += $grade;
            $number_of_grades += 1;
        }
        my $average = $sum / $number_of_grades;
        $output = "$output]  Average: $average \n";
        $total_sum              += $sum;
        $total_number_of_grades += $number_of_grades;
    }
    my $n       = $longest_name_length - length($file_name);
    my $spaces  = ( '-' x $n );
    my $average = $total_sum / $total_number_of_grades;
    return
      "--$spaces$file_name - Average for all students: $average\n$output\n";
}

# Main program

foreach my $i ( 0 .. $#ARGV ) {
    my $file = "$ARGV[$i]";
    print generate_output_files( evaluate_inout_file($file), $file );
}

File: scripting-languages-winter-2017/python/registry.py
#!/usr/bin/env python3
# Witold Bobrowski ; Group 2
#
# Parse text file of students with grades and generate registry

import sys
import argparse as ap
import utils
import ntpath

# CONSTANTS
GRADE, NAME = "grade", "name"

# VARIABLES
longest = {GRADE: 0, NAME: 0}

# FUNCTIONS


def setup_parser():
    parser = ap.ArgumentParser()
    parser.add_argument('files', help="Files to evaluate.",
                        nargs='*', action="store")
    return parser.parse_args()


def evaluate_input_file(file):
    registry = {}
    global longest
    with open(file, 'r') as file:
        for line in file:
            words = line.split()
            name = words[1].lower() + "-" + words[0].lower()
            longest[NAME] = len(str(name)) if len(
                str(name)) > longest[NAME] else longest[NAME]
            grade = utils.get_grade(words[2])
            if grade == 0:
                utils.eprint(
                    "WARNING! \"" + words[2] + "\" is not a valid grade, therefore it will be ignored\n")
            else:
                longest[GRADE] = len(str(grade)) if len(
                    str(grade)) > longest[GRADE] else longest[GRADE]
                registry.setdefault(name, []).append(grade)
    return registry


def format_string(string, separator, longest):
    n = longest - len(str(string))
    spaces = separator * n
    return spaces + str(string)


def generate_output_from(registry, file_name):
    output = ""
    total_sum, total_grades_count = 0, 0
    for key in sorted(registry):
        sum, grades_count = 0, 0
        names = key.split("-")
        output += "    " + format_string(names[0].capitalize() + " " +
                                names[1].capitalize(), " ", longest[NAME]) + "    [ "
        for grade in registry[key]:
            sum += float(grade)
            grades_count += 1
            output += format_string(grade, " ", longest[GRADE]) + ", "
        output += "] Average: " + str(sum / grades_count) + "\n"
        total_sum += sum
        total_grades_count += grades_count
    header = format_string(file_name, "-", longest[NAME]) + \
        " - Average for all students: " + \
        str(total_sum / total_grades_count) + "\n"
    return header + output


# Main
if __name__ == "__main__":
    arguments = setup_parser()
    for file in arguments.files:
        registry = evaluate_input_file(file)
        print(generate_output_from(registry, ntpath.basename(file)))
